File: starter_coding/hello_world.rb
# Import required libraries
require 'net/http'  # For making HTTP requests
require 'uri'       # For parsing URIs

# Check if URL was provided as command line argument
if ARGV.length < 1
  # Print usage message if no URL was provided
  puts "Usage: ruby hello_world.rb <url>"
  # Exit program with error code 1
  exit 1
end

# Get URL from first command line argument
url_string = ARGV[0]

begin
  # Parse the URL string into a URI object
  uri = URI.parse(url_string)
  
  # Create a new HTTP object using the host and port from the URI
  http = Net::HTTP.new(uri.host, uri.port)
  
  # Use HTTPS if the scheme is https
  http.use_ssl = (uri.scheme == 'https')
  
  # Set timeouts to 5 seconds
  http.open_timeout = 5
  http.read_timeout = 5
  
  # Make a GET request to the URI path
  response = http.get(uri.request_uri)
  
  # Print the HTTP status code
  puts "Status Code: #{response.code}"
  
  # Print the first 100 characters of the response body
  puts "Response Content: #{response.body[0...100]}..."
  
rescue => e
  # Print error message if request failed
  puts "Error making request: #{e.message}"
  # Exit program with error code 1
  exit 1
end

File: starter_coding/hello_world.js
// Import required modules
const https = require('https');  // For making HTTPS requests
const http = require('http');   // For making HTTP requests
const url = require('url');    // For parsing URLs

// Function to make an HTTP request
function makeRequest(urlString) {
    // Parse the provided URL
    const parsedUrl = url.parse(urlString);
    
    // Choose the appropriate protocol (http or https)
    const protocol = parsedUrl.protocol === 'https:' ? https : http;
    
    // Return a new Promise to handle the asynchronous request
    return new Promise((resolve, reject) => {
        // Log that request is being made
        console.log(`Making request to: ${urlString}`);
        
        // Make the HTTP/HTTPS request
        const req = protocol.get(urlString, (res) => {
            // Variable to store the response data
            let data = '';
            
            // Log the status code
            console.log(`Status Code: ${res.statusCode}`);
            
            // Event handler for receiving data chunks
            res.on('data', (chunk) => {
                // Append the chunk to our data variable
                data += chunk;
            });
            
            // Event handler for when the response is complete
            res.on('end', () => {
                // Resolve the promise with the collected data
                resolve(data);
            });
        });
        
        // Event handler for request errors
        req.on('error', (err) => {
            // Reject the promise with the error
            reject(`Error making request: ${err.message}`);
        });
        
        // End the request
        req.end();
    });
}

// Main function
async function main() {
    // Check if URL was provided as command line argument
    if (process.argv.length < 3) {
        // Print usage message if no URL was provided
        console.log('Usage: node hello_world.js <url>');
        // Exit program with error code 1
        process.exit(1);
    }
    
    // Get URL from command line argument
    const url = process.argv[2];
    
    try {
        // Make the request and await the response
        const response = await makeRequest(url);
        // Print the first 100 characters of the response
        console.log(`Response Content: ${response.substring(0, 100)}...`);
    } catch (error) {
        // Print any errors that occurred
        console.error(error);
    }
}

// Call the main function
main();
